File: packages/services/Car/tests/libvehiclenetwork-java-test/src/com/android/car/vehiclenetwork/libtest/VehicleNetworkTest.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.car.vehiclenetwork.libtest;

import android.os.HandlerThread;
import android.os.SystemClock;
import android.test.AndroidTestCase;
import android.test.suitebuilder.annotation.MediumTest;
import android.util.ArraySet;
import android.util.Log;

import com.android.car.vehiclenetwork.VehicleNetwork;
import com.android.car.vehiclenetwork.VehicleNetwork.VehicleNetworkListener;
import com.android.car.vehiclenetwork.VehicleNetworkConsts;
import com.android.car.vehiclenetwork.VehicleNetworkConsts.VehiclePropAccess;
import com.android.car.vehiclenetwork.VehicleNetworkConsts.VehiclePropChangeMode;
import com.android.car.vehiclenetwork.VehicleNetworkConsts.VehicleValueType;
import com.android.car.vehiclenetwork.VehicleNetworkProto.VehiclePropConfig;
import com.android.car.vehiclenetwork.VehicleNetworkProto.VehiclePropConfigs;
import com.android.car.vehiclenetwork.VehicleNetworkProto.VehiclePropValue;
import com.android.car.vehiclenetwork.VehicleNetworkProto.VehiclePropValues;
import com.android.car.vehiclenetwork.VehicleNetworkProtoUtil;

import java.util.LinkedList;

@MediumTest
public class VehicleNetworkTest extends AndroidTestCase {
    private static final String TAG = VehicleNetworkTest.class.getSimpleName();
    private static final long TIMEOUT_MS = 2000;

    private final HandlerThread mHandlerThread = new HandlerThread(
            VehicleNetworkTest.class.getSimpleName());
    private VehicleNetwork mVehicleNetwork;
    private final TestListener mListener = new TestListener();

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        mHandlerThread.start();
        mVehicleNetwork = VehicleNetwork.createVehicleNetwork(mListener,
                mHandlerThread.getLooper());
    }

    @Override
    protected void tearDown() throws Exception {
        super.tearDown();
        mHandlerThread.quit();
    }

    public void testListProperties() {
        VehiclePropConfigs configs = mVehicleNetwork.listProperties();
        assertNotNull(configs);
        assertTrue(configs.getConfigsCount() > 0);
        Log.i(TAG, "got configs:" + configs.getConfigsCount());
        for (VehiclePropConfig config : configs.getConfigsList()) {
            Log.i(TAG, VehicleNetworkProtoUtil.VehiclePropConfigToString(config));
        }

        int oneProperty = configs.getConfigs(0).getProp();
        VehiclePropConfigs configs2 = mVehicleNetwork.listProperties(oneProperty);
        assertEquals(1, configs2.getConfigsCount());
        assertTrue(VehicleNetworkProtoUtil.VehiclePropConfigEquals(
                configs.getConfigs(0), configs2.getConfigs(0)));

        VehiclePropConfigs configs3 = mVehicleNetwork.listProperties(-1);
        assertNull(configs3);
    }

    public void testGetProperty() {
        try {
            VehiclePropValue value = mVehicleNetwork.getProperty(-1);
            fail();
        } catch (IllegalArgumentException e) {
            // expected
        }
        VehiclePropConfigs configs = mVehicleNetwork.listProperties();
        assertNotNull(configs);
        assertTrue(configs.getConfigsCount() > 0);
        Log.i(TAG, "got configs:" + configs.getConfigsCount());
        for (VehiclePropConfig config : configs.getConfigsList()) {
            if ((config.getAccess() & VehiclePropAccess.VEHICLE_PROP_ACCESS_READ) != 0) {
                if (config.getProp() == VehicleNetworkConsts.VEHICLE_PROPERTY_RADIO_PRESET) {
                    continue;
                }
                if (config.getProp() >= VehicleNetworkConsts.VEHICLE_PROPERTY_INTERNAL_START &&
                        config.getProp() <= VehicleNetworkConsts.VEHICLE_PROPERTY_INTERNAL_END) {
                    // internal property requires write to read
                    VehiclePropValue v = VehicleNetworkTestUtil.createDummyValue(config.getProp(),
                            config.getValueType());
                    mVehicleNetwork.setProperty(v);
                }
                VehiclePropValue value = mVehicleNetwork.getProperty(config.getProp());
                assertEquals(config.getProp(), value.getProp());
                assertEquals(config.getValueType(), value.getValueType());
                Log.i(TAG, " got property:" +
                        VehicleNetworkProtoUtil.VehiclePropValueToString(value));
            }
        }
    }

    public void testSetProperty() {
        try {
            VehiclePropValue value = VehiclePropValue.newBuilder().
                setProp(-1).
                setValueType(VehicleValueType.VEHICLE_VALUE_TYPE_INT32).
                addInt32Values(0).
                build();
            mVehicleNetwork.setProperty(value);
            fail();
        } catch (SecurityException e) {
            // expected
        }
        VehiclePropConfigs configs = mVehicleNetwork.listProperties();
        assertNotNull(configs);
        assertTrue(configs.getConfigsCount() > 0);
        Log.i(TAG, "got configs:" + configs.getConfigsCount());
        for (VehiclePropConfig config : configs.getConfigsList()) {
            if ((config.getAccess() & VehiclePropAccess.VEHICLE_PROP_ACCESS_WRITE) != 0) {
                if (config.getValueType() == VehicleValueType.VEHICLE_VALUE_TYPE_INT32) {
                    VehiclePropValue value = VehiclePropValue.newBuilder().
                            setProp(config.getProp()).
                            setValueType(config.getValueType()).
                            addInt32Values(0).
                            build();
                    mVehicleNetwork.setProperty(value);
                }
            } else {
                try {
                    VehiclePropValue value = VehiclePropValue.newBuilder().
                            setProp(config.getProp()).
                            setValueType(config.getValueType()).
                            build();
                    mVehicleNetwork.setProperty(value);
                    fail();
                } catch (IllegalArgumentException e) {
                    // expected
                }
            }
        }
    }

    public void testSubscribe() throws Exception {
        try {
            mVehicleNetwork.subscribe(-1, 0);
            fail();
        } catch (SecurityException e) {
            //expected
        }
        VehiclePropConfigs configs = mVehicleNetwork.listProperties();
        assertNotNull(configs);
        assertTrue(configs.getConfigsCount() > 0);
        Log.i(TAG, "got configs:" + configs.getConfigsCount());
        LinkedList<Integer> subscribedProperties = new LinkedList<Integer>();
        for (VehiclePropConfig config : configs.getConfigsList()) {
            if (config.getChangeMode() == VehiclePropChangeMode.VEHICLE_PROP_CHANGE_MODE_STATIC) {
                // cannot subscribe
                try {
                    mVehicleNetwork.subscribe(config.getProp(), config.getSampleRateMin());
                    fail();
                } catch (IllegalArgumentException e) {
                    //expected
                }
            } else if ((config.getAccess() & VehiclePropAccess.VEHICLE_PROP_ACCESS_READ) != 0){
                mVehicleNetwork.subscribe(config.getProp(), config.getSampleRateMin());
                subscribedProperties.add(config.getProp());
                if (config.getProp() >= VehicleNetworkConsts.VEHICLE_PROPERTY_INTERNAL_START &&
                        config.getProp() <= VehicleNetworkConsts.VEHICLE_PROPERTY_INTERNAL_END) {
                    // internal property requires write to get notification
                    VehiclePropValue v = VehicleNetworkTestUtil.createDummyValue(config.getProp(),
                            config.getValueType());
                    mVehicleNetwork.setProperty(v);
                }
            }
        }
        // now confirm event
        for (Integer prop : subscribedProperties) {
            mListener.waitForEvent(prop, TIMEOUT_MS);
            mVehicleNetwork.unsubscribe(prop);
        }
        // wait for already patched events to go away.
        Thread.sleep(1000);
        mListener.resetEventRecord();
        Thread.sleep(2000);
        assertEquals(0, mListener.getActiveEventsCount());
    }

    private class TestListener implements VehicleNetworkListener {
        private final ArraySet<Integer> mEventRecord = new ArraySet<Integer>();

        @Override
        public void onVehicleNetworkEvents(VehiclePropValues values) {
            for (VehiclePropValue value : values.getValuesList()) {
                Log.i(TAG, "event " + VehicleNetworkProtoUtil.VehiclePropValueToString(value));
                synchronized (this) {
                    mEventRecord.add(value.getProp());
                    notifyAll();
                }
            }
        }

        @Override
        public void onHalError(int errorCode, int property, int operation) {
            // TODO Auto-generated method stub
        }

        @Override
        public void onHalRestart(boolean inMocking) {
            // TODO Auto-generated method stub
        }

        private synchronized boolean waitForEvent(Integer prop, long timeoutMs)
                throws InterruptedException {
            long now = SystemClock.elapsedRealtime();
            long end = now + timeoutMs;
            long timeToWait = end - now;
            while (timeToWait > 0 && !mEventRecord.contains(prop)) {
                wait(timeToWait);
                timeToWait = end - SystemClock.elapsedRealtime();
            }
            return mEventRecord.contains(prop);
        }

        private synchronized void resetEventRecord() {
            mEventRecord.clear();
        }

        private synchronized int getActiveEventsCount() {
            return mEventRecord.size();
        }
    }
}

File: packages/services/Car/tests/libvehiclenetwork-native-test/VehicleNetworkTest.cpp
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <unistd.h>

#include <gtest/gtest.h>
#include <binder/IServiceManager.h>
#include <binder/ProcessState.h>
#include <utils/threads.h>
#include <utils/KeyedVector.h>
#include <utils/String8.h>
#include <utils/SystemClock.h>
#include <VehicleNetwork.h>

#include "VehicleNetworkTestListener.h"

namespace android {

// Be careful with name conflict with other tests!. It can lead into wrong virtual function table
// , leading into mysterious crash. Always add test name in front for any class name.
class VehicleNetworkTest : public testing::Test {
public:
    VehicleNetworkTest() :
        mVN(NULL),
        mListener(new VehicleNetworkTestListener()) {
        String8 msg;
        msg.appendFormat("Creating VehicleNetworkTest 0x%p %p %p\n", this, mVN.get(),
                mListener.get());
        std::cout<<msg.string();
    }

    virtual ~VehicleNetworkTest() { }

    sp<VehicleNetwork> getDefaultVN() {
        return mVN;
    }

    VehicleNetworkTestListener& getTestListener() {
        return *mListener.get();
    }

protected:
    void SetUp() {
        String8 msg;
        msg.appendFormat("setUp starts %p %p %p\n", this, mVN.get(),
                mListener.get());
        std::cout<<msg.string();
        ASSERT_TRUE(mListener.get() != NULL);
        sp<VehicleNetworkListener> listener(mListener.get());
        mVN = VehicleNetwork::createVehicleNetwork(listener);
        ASSERT_TRUE(mVN.get() != NULL);
        std::cout<<"setUp ends\n";
    }

protected:
    sp<VehicleNetwork> mVN;
    sp<VehicleNetworkTestListener> mListener;
};


TEST_F(VehicleNetworkTest, listProperties) {
    sp<VehicleNetwork> vn = getDefaultVN();
    sp<VehiclePropertiesHolder> properties = vn->listProperties();
    ASSERT_TRUE(properties.get() != NULL);
    int32_t numConfigs = properties->getList().size();
    ASSERT_TRUE(numConfigs > 0);
    for (auto& config : properties->getList()) {
        String8 msg = String8::format("prop 0x%x\n", config->prop);
        std::cout<<msg.string();
    }
    sp<VehiclePropertiesHolder> propertiesIvalid  = vn->listProperties(-1); // no such property
    ASSERT_TRUE(propertiesIvalid.get() == NULL);
    for (auto& config : properties->getList()) {
        String8 msg = String8::format("query single prop 0x%x\n", config->prop);
        std::cout<<msg.string();
        sp<VehiclePropertiesHolder> singleProperty = vn->listProperties(config->prop);
        ASSERT_EQ((size_t) 1, singleProperty->getList().size());
        vehicle_prop_config_t const * newConfig = *singleProperty->getList().begin();
        ASSERT_EQ(config->prop, newConfig->prop);
        ASSERT_EQ(config->access, newConfig->access);
        ASSERT_EQ(config->change_mode, newConfig->change_mode);
        //TODO add more check
    }
}

TEST_F(VehicleNetworkTest, getProperty) {
    sp<VehicleNetwork> vn = getDefaultVN();
    sp<VehiclePropertiesHolder> properties = vn->listProperties();
    ASSERT_TRUE(properties.get() != NULL);
    int32_t numConfigs = properties->getList().size();
    ASSERT_TRUE(numConfigs > 0);
    for (auto& config : properties->getList()) {
        if (config->prop == VEHICLE_PROPERTY_RADIO_PRESET) {
            continue;
        }
        String8 msg = String8::format("getting prop 0x%x\n", config->prop);
        std::cout<<msg.string();
        ScopedVehiclePropValue value;
        value.value.prop = config->prop;
        value.value.value_type = config->value_type;
        status_t r = vn->getProperty(&value.value);
        if ((config->access & VEHICLE_PROP_ACCESS_READ) == 0) { // cannot read
            ASSERT_TRUE(r != NO_ERROR);
        } else {
            ASSERT_EQ(NO_ERROR, r);
            ASSERT_EQ(config->value_type, value.value.value_type);
        }
    }
}

//TODO change this test to to safe write
TEST_F(VehicleNetworkTest, setProperty) {
    sp<VehicleNetwork> vn = getDefaultVN();
    sp<VehiclePropertiesHolder> properties = vn->listProperties();
    ASSERT_TRUE(properties.get() != NULL);
    int32_t numConfigs = properties->getList().size();
    ASSERT_TRUE(numConfigs > 0);
    for (auto& config : properties->getList()) {
        if (config->prop == VEHICLE_PROPERTY_RADIO_PRESET) {
            continue;
        }
        String8 msg = String8::format("setting prop 0x%x\n", config->prop);
        std::cout<<msg.string();
        ScopedVehiclePropValue value;
        value.value.prop = config->prop;
        value.value.value_type = config->value_type;
        status_t r = vn->setProperty(value.value);
        if ((config->access & VEHICLE_PROP_ACCESS_WRITE) == 0) { // cannot write
            ASSERT_TRUE(r != NO_ERROR);
        } else {
            ASSERT_EQ(NO_ERROR, r);
        }
    }
}

TEST_F(VehicleNetworkTest, setSubscribe) {
    sp<VehicleNetwork> vn = getDefaultVN();
    sp<VehiclePropertiesHolder> properties = vn->listProperties();
    ASSERT_TRUE(properties.get() != NULL);
    int32_t numConfigs = properties->getList().size();
    ASSERT_TRUE(numConfigs > 0);
    for (auto& config : properties->getList()) {
        String8 msg = String8::format("subscribing property 0x%x\n", config->prop);
        std::cout<<msg.string();
        status_t r = vn->subscribe(config->prop, config->max_sample_rate);
        if (((config->access & VEHICLE_PROP_ACCESS_READ) == 0) ||
                (config->change_mode == VEHICLE_PROP_CHANGE_MODE_STATIC)) { // cannot subsctibe
            ASSERT_TRUE(r != NO_ERROR);
        } else {
            if ((config->prop >= (int32_t)VEHICLE_PROPERTY_INTERNAL_START) &&
                    (config->prop <= (int32_t)VEHICLE_PROPERTY_INTERNAL_END)) {
                // internal property requires write for event notification.
                ScopedVehiclePropValue value;
                value.value.prop = config->prop;
                value.value.value_type = config->value_type;
                status_t r = vn->setProperty(value.value);
                ASSERT_EQ(NO_ERROR, r);
            }
            ASSERT_EQ(NO_ERROR, r);
            ASSERT_TRUE(getTestListener().waitForEvent(config->prop, 0, 2000000000));
        }
    }
    for (auto& config : properties->getList()) {
        vn->unsubscribe(config->prop);
    }
    usleep(1000000);
    //TODO improve this as this will wait for too long
    for (auto& config : properties->getList()) {
        int initialCount = getTestListener().getEventCount(config->prop);
        ASSERT_TRUE(!getTestListener().waitForEvent(config->prop, initialCount, 1000000000));
    }
}

}; // namespace android
